src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -528,9 +751,50 @@ PayChanClaim::doApply()
         }
 
         (*slep)[sfBalance] = ctx_.tx[sfBalance];
-        XRPAmount const reqDelta = reqBalance - chanBalance;
+        STAmount const reqDelta = reqBalance - chanBalance;
         assert(reqDelta >= beast::zero);
-        (*sled)[sfBalance] = (*sled)[sfBalance] + reqDelta;
+        if (isXRP(reqDelta))
+            (*sled)[sfBalance] = (*sled)[sfBalance] + reqDelta;
+        else
+        {
+            // xfer locked tokens to satisfy claim
+            // RH NOTE: there's no ledger modification before this point so
+            // no reason to do a dry run first
+            if (!ctx_.view().rules().enabled(featurePaychanAndEscrowForTokens))
+                return temDISABLED;
+
+            // compute transfer fee, if any
+            Rate lockedRate = ripple::Rate(slep->getFieldU32(sfTransferRate));
```

Comment:
I think we need to check for optional since sfTransferRate is now an optional field.

would this like this work?

```
auto const transferRate = slep->at(~sfTransferRate).value_or(0u);
if (!transferRate)
    return tecINTERNAL;

Rate lockedRate = ripple::Rate(transferRate);
```

Reply:
The value of the transfer rate if it does not exist is 1. The `sfTransferRate` is optional but once this code goes into effect the transfer rate will be 1 even if there exists no transfer rate and the rate will always be added to the line for any issued currency or iou. The above code will result in 1 being the transfer rate (if no rate exists).

I can write it how you did above but first wanted to explain.

Reply:
Sorry I was incorrect. `transferRate(...)` results in 1 if no rate exists for the iou. I can guard the optional.

Reply:
yeah i think as long as it's guard then it's fine

Reply:
`slep->isFieldPresent(sfTransferRate);` should be used here then